### PR TITLE
Add bringyourownhost provider ytt template and configuration

### LINKED ISSUE #823: Enable BYOH provider in Tanzu-framework
(This is used to request new product features)

**Describe the feature request**
A user would like to be able to create, list, delete and scale out/in a BYOH workload cluster with Tanzu CLI.

- [ ] APIs
- [] Addons
- [X] CLI
- [ ] Docs
- [] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [X] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by 4everming**
### What this PR does / why we need it
To enable BYOH provider in tanzu-framework
In this PR some YTT templates and variables to are defined.

### Which issue(s) this PR fixes
Fixes #823

### Release note

```release-note

```

### PR Checklist

- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+kind: KubeadmControlPlane
+metadata:
+  name: '${ CLUSTER_NAME }-control-plane'
+  namespace: '${ NAMESPACE }'
+spec:
+  kubeadmConfigSpec:
+    useExperimentalRetryJoin: true
+    clusterConfiguration:
+      apiServer:
+        timeoutForControlPlane: "8m0s"
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:55]: Does 8m is enough for most of the cases?  Where does this number come from?

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+          criSocket: /var/run/containerd/containerd.sock
+          kubeletExtraArgs:
+            tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+          name: '{{ ds.meta_data.hostname }}'
+      preKubeadmCommands:
+      - hostname "{{ ds.meta_data.hostname }}"
+      - echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
+      - echo "127.0.0.1   localhost" >>/etc/hosts
+      - echo "127.0.0.1   {{ ds.meta_data.hostname }}" >>/etc/hosts
+      - echo "{{ ds.meta_data.hostname }}" >/etc/hostname
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:152]:
```suggestion
```

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        kubeletExtraArgs:
+          tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+    joinConfiguration:
+      nodeRegistration:
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs:
+          tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+    preKubeadmCommands:
+    - echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
+    - echo "127.0.0.1   localhost" >>/etc/hosts
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:123]:
```suggestion

```

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            - start
+            env:
+            - name: vip_arp
+              value: "true"
+            - name: vip_leaderelection
+              value: "true"
+            - name: address
+              value: '${ VSPHERE_CONTROL_PLANE_ENDPOINT }'
+            - name: vip_interface
+              value: '${ VIP_NETWORK_INTERFACE }'
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:83]:
```suggestion
              value: {{ .DefaultNetworkName }}
```

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/overlay.yaml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    files:
+    #@ if not data.values.AVI_CONTROL_PLANE_HA_PROVIDER:
+    #@overlay/match by=overlay.index(0)
+    - content: #@ yaml.encode(kube_vip_pod())
+    #@ else:
+    #@overlay/match by=overlay.index(0)  
+    #@overlay/remove
+    - content:
+    #@ end
+    users:
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/overlay.yaml:146]: BYOH don't need this section.   Please update this file according to the cluster template file.

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml (hunk 1 above)
Comment (4everming) [reply pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:55]: For "infrastructure-aws" and "infrastructure-vsphere", they both use this value as the timeout value. For BYOH, either remove this configuration or consider a new value for this should be OK.

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        kubeletExtraArgs:
+          cgroup-driver: cgroupfs
+          eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
+  machineTemplate:
+    infrastructureRef:
+      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+      kind: ByoMachineTemplate
+      name: byohcluster
+  replicas: ${ CONTROL_PLANE_MACHINE_COUNT }
+  version: '${ KUBERNETES_VERSION }'
```
Comment (4everming) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:130]: Already use this config as the Kubernetes version for Byoh cluster, does it still need to add this?
"     kubernetesVersion: '${ KUBERNETES_VERSION }' "

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/infrastructure-components.yaml [hunk 7]
```diff
@@ -0,0 +1,924 @@
...
+      labels:
+        control-plane: controller-manager
+    spec:
+      containers:
+        - args:
+            - --metrics-addr=127.0.0.1:8080
+            - --enable-leader-election
+          command:
+            - /manager
+          image: gcr.io/k8s-staging-cluster-api/caph-manager:latest
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/infrastructure-components.yaml:839]:
```suggestion
          image: projects.registry.vmware.com/cluster_api_provider_byoh/caph-manager:v0.1.0
```

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+            - start
+            env:
+            - name: vip_arp
+              value: "true"
+            - name: vip_leaderelection
+              value: "true"
+            - name: address
+              value: '${ BYOH_CONTROL_PLANE_ENDPOINT }'
+            - name: vip_interface
+              value:
```
Comment (yixingjia) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:84]:
```suggestion
              value: {{ .DefaultNetworkInterfaceName }}
```

File: pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  name: '${ CLUSTER_NAME }-md-0'
+  namespace: '${ NAMESPACE }'
+spec:
+  template:
+    spec:
+      joinConfiguration:
+        nodeRegistration:
+          criSocket: /var/run/containerd/containerd.sock
+          kubeletExtraArgs:
+            cgroup-driver: cgroupfs
```
Comment (anusha94) [inline pkg/v1/providers/infrastructure-byoh/v0.1.0/ytt/base-template.yaml:144]: @yixingjia 
Given that we will use `systemd` for `containerd`, we should probably stick to the `kubeadm 1.22` **default** of `systemd` and not overwrite with `cgroupfs`. Thoughts?

**Comment by vuil** [comment]
Thanks for the contribution! But based on earlier discussion, I am marking this do-not-merge/hold until we have a new iteration of the implementation.

**Comment by vuil** [comment]
Thanks for the contribution but. Closing until we revisit it again with updated implementation.

**Pull request closed without merging**